### PR TITLE
Use yauzl instead of adm-zip

### PR DESCRIPTION
**Pull request by shinnn**
* I added more tests to cover various cases.
* I replaced [adm-zip](https://github.com/cthackers/adm-zip) with [yauzl](https://github.com/thejoshwolfe/yauzl).
  * yauzl is really faster than adm-zip.
  * [yauzl can read buffer directly](https://github.com/thejoshwolfe/yauzl#frombufferbuffer-callback). So we don't need to create/remove temporary files.
  * [adm-zip project has no test](https://github.com/cthackers/adm-zip/blob/f757ce4c264c042affdd5121625a38208745bf87/test/index.js). It's risky for us to use such an untested module.
* I added benchmark script.
  * Here is the benchmark result on my Mac OS X.

  ```
                  Extract ZIP archive
         1,783 op/s » current version (using yauzl)
           367 op/s » older version (using adm-zip)

                  Extract ZIP archive (nested file)
         1,992 op/s » current version (using yauzl)
           339 op/s » older version (using adm-zip)

                  Extract ZIP archive (multiple files)
         1,784 op/s » current version (using yauzl)
           431 op/s » older version (using adm-zip)
  ```
  As you can see, yauzl version is about 5 times faster than adm-zip version!

Benchmark script is included in the last commit. If you don't want to add it, I'll remove the last commit.

### CODE REVIEW

**Comment by kevva** [comment]
Love it! Yeah, remove the benchmark as it won't work anyway when I release this. Thank you!

File: index.js [hunk 1]
```diff
@@ -30,19 +29,48 @@ module.exports = function (opts) {
...
+			zipFile
+				.on('error', cb)
+				.on('entry', function (entry) {
+					count++;
+
+					if (entry.fileName.charAt(entry.fileName.length - 1) === '/') {
+						if (count === zipFile.entryCount) {
+							cb();
+						}
+						return;
```
Comment (kevva) [inline index.js]:
```js

return
```

File: index.js [hunk 2]
```diff
@@ -30,19 +29,48 @@ module.exports = function (opts) {
...
+
+								self.push(new File({
+									contents: buf,
+									path: stripDirs(entry.fileName, opts.strip)
+								}));
+
+								if (count === zipFile.entryCount) {
+									cb();
+								}
+								done();
```
Comment (kevva) [inline index.js]:
```js
// linebreak
done();
```

File: index.js [hunk 3]
```diff
@@ -30,19 +29,48 @@ module.exports = function (opts) {
...
+						if (err) {
+							cb(err);
+							return;
+						}
 
-			rm(filepath, cb);
+						readStream
+							.on('error', cb)
+							.pipe(new BufferStreams(function (err, buf, done) {
+
```
Comment (kevva) [inline index.js]: Remove this line.

**Comment by kevva** [comment]
Really sick benchmarks. This is awesome!

**Comment by shinnn** [comment]
Yeah, yauzl is awesome. We should be grateful to @thejoshwolfe.

**Comment by kevva** [comment]
:+1:

**Comment by sindresorhus** [comment]
Woah, nice!

**Pull request merged**